pkg/webhook/ipampool/validation/validation.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"strings"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// validator for validating Resource Quota CRD.
+type validator struct {
+	client ctrlruntimeclient.Client
+}
+
+// NewValidator returns a new Resource Quota validator.
+func NewValidator(client ctrlruntimeclient.Client) *validator {
+	return &validator{
+		client: client,
+	}
+}
+
+var _ admission.CustomValidator = &validator{}
+
+func (v *validator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	return v.validate(ctx, obj)
+}
+
+func (v *validator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	if err := v.validate(ctx, newObj); err != nil {
+		return err
+	}
+
+	newIPAMPool := newObj.(*kubermaticv1.IPAMPool)
+	oldIPAMPool := oldObj.(*kubermaticv1.IPAMPool)
+
+	// loop old IPAMPool datacenters
+	for dc, dcOldConfig := range oldIPAMPool.Spec.Datacenters {
+		dcNewConfig, dcExistsInNewPool := newIPAMPool.Spec.Datacenters[dc]
+		if !dcExistsInNewPool {
+			err := v.validateDCRemoval(ctx, oldIPAMPool, dc)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		if dcOldConfig.PoolCIDR != dcNewConfig.PoolCIDR {
+			return errors.New("it's not allowed to update the pool CIDR for a datacenter")
+		}
+
+		if dcOldConfig.Type != dcNewConfig.Type {
+			return errors.New("it's not allowed to update the allocation type for a datacenter")
+		}
+
+		switch dcOldConfig.Type {
+		case kubermaticv1.IPAMPoolAllocationTypeRange:
+			if dcOldConfig.AllocationRange != dcNewConfig.AllocationRange {
+				return errors.New("it's not allowed to update the allocation range for a datacenter")
+			}
+		case kubermaticv1.IPAMPoolAllocationTypePrefix:
+			if dcOldConfig.AllocationPrefix != dcNewConfig.AllocationPrefix {
+				return errors.New("it's not allowed to update the allocation prefix for a datacenter")
+			}
+		}
```

Comment:
Discussing with @rastislavs , we decided to change this requirement because it wouldn't be consistent to have allocations that are not compatible with the updated pool. Also, the controller algorithm is checking all allocations for building a map of already allocated "subnets" from the pool and, specially for the prefix type, it would stop working. For the range type, a small change would make it work, but for now we decided to keep it simple and forbid this change (anyway, the user can still delete the pool and apply it again).

Reply:
ok, makes sense